File: scala/sdktest/GetMerchantContractNumberEntityTest.scala
// Generated basic-flow test for the get_merchant_contract_number entity (model-driven;
// mirrors the java TestEntity generator). A dependency-free scala-cli test
// object driven by SdkEntityTestMain. Runs against the in-memory test
// transport seeded with the shipped GetMerchantContractNumberTestData.json fixtures.

import java.util.{ArrayList, LinkedHashMap, List => JList, Map => JMap}

import voxgig.bluefintecsmerchantservicessdk.core.{Helpers, BluefinTecsMerchantServicesSDK}
import voxgig.bluefintecsmerchantservicessdk.utility.struct.Struct

object GetMerchantContractNumberEntityTest {

  def run(rep: SdkTestReport): Unit = {
    rep.scope("get_merchant_contract_number.instance") {
      val testsdk = BluefinTecsMerchantServicesSDK.testSDK()
      val ent = testsdk.getMerchantContractNumber(null)
      rep.check("get_merchant_contract_number.instance", ent != null, "expected non-null get_merchant_contract_number entity")
    }

    rep.scope("get_merchant_contract_number.basic") {
      val entityData = Helpers.toMapAny(SdkTestSupport.readJson(
          "../.sdk/test/entity/get_merchant_contract_number/GetMerchantContractNumberTestData.json"))
      val options = new LinkedHashMap[String, Object]()
      options.put("entity", entityData.get("existing"))
      val client = BluefinTecsMerchantServicesSDK.testSDK(options, null)

      val idmap = new LinkedHashMap[String, Object]()
      idmap.put("get_merchant_contract_number01", "GET_MERCHANT_CONTRACT_NUMBER01")
      idmap.put("get_merchant_contract_number02", "GET_MERCHANT_CONTRACT_NUMBER02")
      idmap.put("get_merchant_contract_number03", "GET_MERCHANT_CONTRACT_NUMBER03")
      val now = System.currentTimeMillis()

      // CREATE
      val getMerchantContractNumberRef01Ent = client.getMerchantContractNumber(null)
      var getMerchantContractNumberRef01Data = Helpers.toMapAny(Struct.getprop(
          Struct.getpath(entityData, "new.get_merchant_contract_number"), "get_merchant_contract_number_ref01"))
      val getMerchantContractNumberRef01DataResult = getMerchantContractNumberRef01Ent.create(getMerchantContractNumberRef01Data, null)
      getMerchantContractNumberRef01Data = Helpers.toMapAny(getMerchantContractNumberRef01DataResult)
      rep.check("get_merchant_contract_number.create.map", getMerchantContractNumberRef01Data != null, "expected create result to be a map")
    }
  }
}

File: rust/tests/get_merchant_contract_number_entity_test.rs
// Generated basic-flow test for the get_merchant_contract_number entity (model-driven;
// mirrors the go TestEntity generator).

#![allow(unused_variables, unused_mut, unused_imports)]

mod common;

use std::rc::Rc;

use common::*;

use bluefin_tecs_merchant_services_sdk::core::helpers::{getp, getpath, ja, jo, now_ms, setp, to_map};
use bluefin_tecs_merchant_services_sdk::utility::voxgigstruct as vs;
use bluefin_tecs_merchant_services_sdk::{test_sdk, Entity, BluefinTecsMerchantServicesEntity, BluefinTecsMerchantServicesSDK, Value};

#[test]
fn get_merchant_contract_number_entity_instance() {
    let testsdk = test_sdk(Value::Noval, Value::Noval);
    let ent = testsdk.get_merchant_contract_number(Value::Noval);
    assert_eq!(ent.get_name(), "get_merchant_contract_number");
}

#[test]
fn get_merchant_contract_number_entity_basic() {
    let setup = get_merchant_contract_number_basic_setup(Value::Noval);
    // Per-op sdk-test-control.json skip — the basic test exercises a flow
    // with multiple ops; skipping any op skips the whole flow.
    let mode = if setup.live { "live" } else { "unit" };
    for op in ["create"] {
        let (skip, reason) = is_control_skipped("entityOp", &format!("get_merchant_contract_number.{}", op), mode);
        if skip {
            let reason = if reason.is_empty() {
                "skipped via sdk-test-control.json".to_string()
            } else {
                reason
            };
            eprintln!("skip: {}", reason);
            return;
        }
    }
    // The basic flow consumes synthetic IDs from the fixture. In live mode
    // without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only {
        eprintln!("skip: live entity test uses synthetic IDs from fixture — set BLUEFIN_TECS_MERCHANT_SERVICES_TEST_GET_MERCHANT_CONTRACT_NUMBER_ENTID JSON to run live");
        return;
    }
    let client = setup.client.clone();
    // CREATE
    let get_merchant_contract_number_ref01_ent = client.get_merchant_contract_number(Value::Noval);
    let get_merchant_contract_number_ref01_data = to_map(&getp(
        &getpath(&["new", "get_merchant_contract_number"], &setup.data),
        "get_merchant_contract_number_ref01",
    ));

    let get_merchant_contract_number_ref01_data_result = get_merchant_contract_number_ref01_ent
        .create(get_merchant_contract_number_ref01_data.clone(), Value::Noval)
        .expect("create failed");
    let get_merchant_contract_number_ref01_data = to_map(&get_merchant_contract_number_ref01_data_result.data(None));
    assert!(
        matches!(get_merchant_contract_number_ref01_data, Value::Map(_)),
        "expected create result to be a map"
    );

}

fn get_merchant_contract_number_basic_setup(extra: Value) -> EntityTestSetup {
    load_env_local();

    let mut entity_data_file = manifest_dir();
    entity_data_file.push("..");
    entity_data_file.push(".sdk");
    entity_data_file.push("test");
    entity_data_file.push("entity");
    entity_data_file.push("get_merchant_contract_number");
    entity_data_file.push("GetMerchantContractNumberTestData.json");

    let entity_data = read_json(&entity_data_file);

    let options = jo(vec![("entity", getp(&entity_data, "existing"))]);

    let client = test_sdk(options, extra.clone());

    // Generate idmap via transform, matching the TS pattern.
    let idmap = vs::transform(
        &ja(vec![Value::str("get_merchant_contract_number01"), Value::str("get_merchant_contract_number02"), Value::str("get_merchant_contract_number03")]),
        &jo(vec![(
            "`$PACK`",
            ja(vec![
                Value::str(""),
                jo(vec![
                    ("`$KEY`", Value::str("`$COPY`")),
                    (
                        "`$VAL`",
                        ja(vec![
                            Value::str("`$FORMAT`"),
                            Value::str("upper"),
                            Value::str("`$COPY`"),
                        ]),
                    ),
                ]),
            ]),
        )]),
        None,
    )
    .unwrap_or_else(|_| Value::empty_map());

    // Detect ENTID env override before env_override consumes it. When live
    // mode is on without a real override, the basic test runs against
    // synthetic IDs from the fixture and 4xx's.
    let entid_env_raw = std::env::var("BLUEFIN_TECS_MERCHANT_SERVICES_TEST_GET_MERCHANT_CONTRACT_NUMBER_ENTID").unwrap_or_default();
    let idmap_overridden =
        !entid_env_raw.trim().is_empty() && entid_env_raw.trim().starts_with('{');

    let env = env_override(jo(vec![
        ("BLUEFIN_TECS_MERCHANT_SERVICES_TEST_GET_MERCHANT_CONTRACT_NUMBER_ENTID", idmap.clone()),
        ("BLUEFIN_TECS_MERCHANT_SERVICES_TEST_LIVE", Value::str("FALSE")),
        ("BLUEFIN_TECS_MERCHANT_SERVICES_TEST_EXPLAIN", Value::str("FALSE")),
        ("BLUEFIN_TECS_MERCHANT_SERVICES_APIKEY", Value::str("NONE")),
    ]));

    let idmap_resolved = match to_map(&getp(&env, "BLUEFIN_TECS_MERCHANT_SERVICES_TEST_GET_MERCHANT_CONTRACT_NUMBER_ENTID")) {
        Value::Map(m) => Value::Map(m),
        _ => to_map(&idmap),
    };

    let live = getp(&env, "BLUEFIN_TECS_MERCHANT_SERVICES_TEST_LIVE") == Value::str("TRUE");

    let client = if live {
        let merged = vs::merge(
            &ja(vec![jo(vec![("apikey", getp(&env, "BLUEFIN_TECS_MERCHANT_SERVICES_APIKEY"))]), extra]),
            None,
        );
        BluefinTecsMerchantServicesSDK::new(to_map(&merged))
    } else {
        client
    };

    EntityTestSetup {
        client,
        data: entity_data,
        idmap: idmap_resolved,
        env: env.clone(),
        explain: getp(&env, "BLUEFIN_TECS_MERCHANT_SERVICES_TEST_EXPLAIN") == Value::str("TRUE"),
        live,
        synthetic_only: live && !idmap_overridden,
        now: now_ms(),
    }
}
